dhcpoptinj.1
'\" t
.\"
.\" Copyright © 2019 Andreas Misje
.\"
.\" This file is part of dhcpoptinj.
.\"
.\" dhcpoptinj is free software: you can redistribute it and/or modify it under
.\" the terms of the GNU General Public License as published by the Free
.\" Software Foundation, either version 3 of the License, or (at your option)
.\" any later version.
.\"
.\" dhcpoptinj is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
.\" more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with dhcpoptinj. If not, see <http://www.gnu.org/licenses/>.
.\"
.
.Dd February 10, 2019
.Dt DHCPOPTINJ 1 fjas
.Os
.Sh NAME
.Nm dhcpoptinj
.Nd DHCP option injector, a tool to manipulate DHCP packet options
.Sh SYNOPSIS
.Nm
.Bk
.Op Fl df
.Ek
.Bk
.Op Fl Fl forward-on-fail
.Ek
.Bk
.Op Fl i Ns | Ns Fl r
.Ek
.Bk
.Op Fl p Op Ar pid_file
.Ek
.Bk
.Fl q Ar queue_num
.Ek
.Bk
.Fl o Ar dhcp_option
.Ek
.Bk
.Op Po Fl o Ar dhcp_option Pc ...
.Ek
.Bk
.Nm
.Fl h Ns | Ns Fl v
.Ek
.Sh DESCRIPTION
.Nm
is a fairly simple tool for manipulating DHCP options in a BOOTP/DHCP packet.
It relies on netfilter queue and nftables/iptables rules, in which you specify what packets to send to
.Nm .
.Nm
waits for packets to arrive in a netfilter queue.
It will ensure that a packet is in fact a BOOTP/DHCP packet, and if so proceed to inject or replace options.
It will recalculate the IPv4 header checksum, disable the UDP checksum (for a simpler implementation) and then give the packet back to netfilter.
.Pp
Based on the options
.Fl Fl ignore-existing-opt
and
.Fl Fl remove-existing-opt
.Nm
may be configured to replace matching options or leave them in place.
.Fl Fl forward-on-fail
can be used to accept the packet (leaving it unaltered) if the package mangling should fail for any reason.
.Ss Invocation
.Nm
incorrectly requires flags to run.
The netfilter queue number
.Pq Fl Fl queue
is necessary, and also at least one DHCP option
.Pq Fl Fl option .
See
.Sx EXAMPLES .
.Ss Argument list processing
.Nm
is invoked only with options and no arguments:
.Bl -tag -width ".Fl r, .Fl .Fl remove-existing-opt" -offset indent
.It Fl d , Fl Fl debug
Make
.Nm
tell you as much as possible about what it does and tries to do.
.It Fl f , Fl Fl foreground
Prevent
.Nm
from running in the background.
.It "\ \ \ " Fl Fl forward-on-fail
If the process of injecting options should fail, let the unaltered DHCP packet pass through.
The default behaviour is to drop the packet if options could not be injected.
.It Fl h , Fl Fl help
Print a help text.
.It Fl i , Fl Fl ignore-existing-opt
Proceed if an injected option already exists in the original packet.
Unless
.Fl Fl remove-existing-opt
is provided, the default behaviour is to drop the packet.
.It Fl o , Fl Fl option Ar dhcp_option
DHCP option to inject as a hex string, where the first byte indicates the option code.
The option length field is automatically calculated and must be omitted.
Several options may be injected.
.Pp
The hex string may be delimited by non-hexadecimal characters for readability.
The following hex strings all produce the same result: '01 02 03', '01:02:03', '010203', '01 <-> 02 <-> 03'.
Remember to quote these arguments if they contains spaces or other characters that have special meaning to your shell.
.It Fl p , Fl Fl pid-file Op Pa pid_file
Write PID to file, using specified path or a default sensible location.
.It Fl q , Fl Fl queue Ar queue_num
The netfilter queue number to use
.It Fl r , Fl Fl remove-existing-opt
Remove existing DHCP options of the same kind as those to be injected.
.It Fl v , Fl Fl version
Display version.
.El
.Pp
.Ss DHCP options
All the DHCP options specified with the
.Fl o Ns / Ns Fl Fl option
flag will be added before the terminating option (end option, 255).
The packet is padded if necessary and sent back to netfilter.
The IPv4 header checksum is recalculated, but the UDP checksum is set to 0 (disabled).
None of the added options are checked for whether they are valid, or whether the option codes are valid.
Options are currently not (automatically) padded individually, but they can be manually padded by adding options with code 0 (one pad byte per option).
This special option is the only option that does not have any payload (the end option, 255, is inserted automatically and cannot be manually added).
Padding individual options should not be necessary.
.Pp
The option hex string is written as a series of two-digit pairs,
optionally delimited by one or more non-hexadecimal characters: '466A6173','46 6A 61 73', '46:6A:61:73' etc.
There is a maximum limit of 256 bytes per option, excluding the option code (the first byte) and the automatically inserted length byte.
At least one option must be provided.
.Pp
If the packet already contains a DHCP option that is to be injected (matched by code), the behaviour depends on the command line options
.Fl Fl ignore-existing-opt
and
.Fl Fl remove-existing-opt :
.Bl -tag -width ".Pq none" -offset indent
.It Pq none
The packet will be dropped.
.It Fl i
The existing options are ignored and the injected options are added.
.It Fl r
Any existing options are removed and the injected options are added.
.El
.Pp
.Em Note
that injected options will not be injected in the same place as those that may have been removed if using
.Fl r .
However, this should not matter.
.Sh EXIT STATUS
.Ex -std dhcpoptinj
.Sh EXAMPLES
Let us say you have two interfaces bridged together, eth0 and eth1.
Let us say you want to intercept all BOOTP requests coming from eth0 and inject the relay agent information option (82/0x52).
Let us make up a silly payload: An agent circuit ID sub-option with the value "Fjas".
.Pp
Add a rule to the iptables mangle table:
.Pp
.Dl sudo iptables -t mangle -A PREROUTING -m physdev --physdev-in eth0 -p udp --dport 67 -j NFQUEUE --queue-num 42
.Pp
Then run
.Nm
(let us run it in the foreground with extra debug output):
.Pp
.Dl sudo dhcpoptinj -d -f -q 42 -o'52 01 04 46 6A 61 73'
.Pp
.Em Note
that
.Nm
must be run by a user with the CAP_NET_ADMIN capability.
You do not need to, and you really should not run
.Nm
as root.
Instead, you can for instance grant the CAP_NET_ADMIN capability to the binary (using
.Em setcap )
and limit execution rights to only a specific user or group.
This is a method used for running Wireshark as non-root, so you will find several guides helping you accomplish this.
.Pp
Now send a DHCP packet to the eth0 interface and watch it (using a tool like Wireshark
.Pq Lk https://\:www.wireshark.org/
having been modified when it reaches the bridged interface.
It should have the injected option at the end of the option list.
If you capture the incoming DHCP packet with Wireshark, it will appear unmodified although it will in fact be mangled.
.Pp
Note the format of the argument to the
.Fl o
option: It should be a hexadecimal string starting with the DHCP option code followed by the option payload.
The option length (the byte that normally follows the option code) is automatically calculated and must not be specified.
The hex string can be delimited by non-hexadecimal characters for readability.
All options must have a payload, except for the special pad option
.Pq Lk https://\:tools.ietf.org/\:html/\:rfc2132#section-2
(code 0).
.Pp
The layout of the nonsensical option used in this example, 52 01 04 46 6A 61 73, (first the DHCP option layout
.Pq Lk https://\:tools.ietf.org/\:html/\:rfc2132#section-2 ,
then the specific relay agent information option sub-option layout
.Pq Lk https://tools.ietf.org/html/rfc3046#section-2.0 )
is as follows:
.Pp
.TS
allbox tab(;);
c c c
c c c.
Code;Length;Data
52;(auto);01 04 46 6A 61 73 ("Fjas")
.TE
.Pp
.TS
allbox tab(;);
c c c
c c c.
Sub-opt.;Lenth;Data
01;4;46 6A 61 73 ("Fjas")
.TE
.Sh SEE ALSO
iptables(8), nftables(8)
.Sh AUTHORS
.An "Andreas Misje" Aq amisje@gmail.com


src/dhcpoptinj.c
/* 
 * Copyright © 2015 Andreas Misje
 *
 * This file is part of dhcpoptinj.
 *
 * dhcpoptinj is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.  
 *
 * dhcpoptinj is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with dhcpoptinj. If not, see <http://www.gnu.org/licenses/>.
 */

#include <sys/types.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <linux/netfilter.h>
#include <libnetfilter_queue/libnetfilter_queue.h>
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include <stdbool.h>
#include <syslog.h>
#include <unistd.h>
#include "config.h"
#include <stdarg.h>
#include <signal.h>
#include <errno.h>
#include <sys/stat.h>
#include <libgen.h>
#include "ipv4.h"
#include "udp.h"
#include "dhcp.h"

#define MIN_BOOTP_SIZE 300

struct Packet
{
	struct IPv4Header ipHeader;
	struct UDPHeader udpHeader;
	struct BootP bootp;
} __attribute__((packed));

enum MangleResult
{
	Mangle_OK = 0,
	Mangle_mallocFail,
	Mangle_optExists,
};

/* Somewhat arbitrary, feel free to change */
static const int maxPacketSize = 2048;
/* The netfilter queue length 20 is also arbitrary. Hopefully it is
 * sufficient. */
static const int maxQueueLen = 20;
static struct Config *config;
static bool daemonised;
static sig_atomic_t escapeMainLoop;
static sig_atomic_t signalCaught;

static int inspectPacket(struct nfq_q_handle *queue, struct nfgenmsg *pktInfo, 
		struct nfq_data *pktData, void *userData);
static bool packetIsComplete(const uint8_t *data, size_t size);
static bool packetIsDHCP(const uint8_t *data);
/* Inject DHCP options into DHCP packet */
static enum MangleResult manglePacket(const uint8_t *origData, size_t origDataSize,
		uint8_t **newData, size_t *newDataSize);
static enum MangleResult mangleOptions(const uint8_t *origData, size_t origDataSize,
		uint8_t *newData, size_t *newDataSize);
/* Write a message to syslog or standard stream, depending on whether the
 * process is run as a daemon or not */
static void logMessage(int priority, const char *format, ...);
static void simplifyProgramName(char *programName);
static void writePID(void);
static void removePIDFile(void);
static void destroyConfig(void);
static void initSignalHandler(void);
static void setEscapeMainLoopFlag(int signal);
static void initLog(const char *programName);
/* Debug-print all options to inject */
static void debugLogOptions(void);
/* Very simple check of the provided option codes, warning user if something
 * looks incorrect */
static void inspectOptions(void);
/* Debug-print packet header */
static void debugLogPacketHeader(const uint8_t *data, size_t size);
/* Debug-print packet's existing DHCP options */
static void debugLogOption(const struct DHCPOption *option);

int main(int argc, char *argv[])
{
	simplifyProgramName(argv[0]);
	config = conf_parseOpts(argc, argv);
	initLog(argv[0]);

	debugLogOptions();
	inspectOptions();

	logMessage(LOG_DEBUG, "Initialising netfilter queue …\n");

	struct nfq_handle *nfq = nfq_open();
	if (!nfq)
	{
		/* Most likely causes are insufficient permissions (missing
		 * CAP_NET_ADMIN capability) or an another process already bound to the
		 * same queue. */
		logMessage(LOG_ERR, "Failed to initialise netfilter queue library: %s\n",
				strerror(errno));
		exit(EXIT_FAILURE);
	}
	nfq_unbind_pf(nfq, AF_INET);
	if (nfq_bind_pf(nfq, AF_INET) < 0)
	{
		logMessage(LOG_ERR, "Failed to bind queue handler to AF_INET: %s\n",
				strerror(errno));
		exit(EXIT_FAILURE);
	}

	struct nfq_q_handle *queue = nfq_create_queue(nfq, config->queue, &inspectPacket, 
			NULL);
	if (!queue)
	{
		logMessage(LOG_ERR, "Failed to create netfilter queue for queue %d: %s\n",
				config->queue, strerror(errno));
		exit(EXIT_FAILURE);
	}

	if (nfq_set_mode(queue, NFQNL_COPY_PACKET, maxPacketSize) < 0)
	{
		logMessage(LOG_ERR, "Failed to set netfilter queue mode: %s\n", strerror(
					errno));
		exit(EXIT_FAILURE);
	}

	if (nfq_set_queue_maxlen(queue, maxQueueLen) < 0)
	{
		logMessage(LOG_ERR, "Failed to set netfilter queue length: %s\n", strerror(
					errno));
		exit(EXIT_FAILURE);
	}

	if (!config->foreground)
	{
		logMessage(LOG_DEBUG, "Daemonising …\n");
		if (daemon(false, false))
		{
			logMessage(LOG_ERR, "Failed to daemonise: daemon() failed: %s\n", 
					strerror(errno));
			exit(EXIT_FAILURE);
		}
		umask(022);
		daemonised = true;
	}
	writePID();

	initSignalHandler();

	if (config->debug)
		logMessage(LOG_DEBUG, "Initialisation completed. Waiting for packets to mangle …\n");
	else
		logMessage(LOG_INFO, "Started\n");

	int exitCode = EXIT_SUCCESS;
	int queueFd = nfq_fd(nfq);
	for (; !escapeMainLoop; )
	{
		char packet[maxPacketSize] __attribute__((aligned));
		ssize_t bytes = recv(queueFd, packet, sizeof(packet), 0);
		if (bytes < -1)
		{
			logMessage(LOG_ERR, "Failed to retrieve packet: %s\n", strerror(errno));
			exitCode = EXIT_FAILURE;
			break;
		}
		else if (bytes > 0)
		{
			logMessage(LOG_DEBUG, "Received %zd bytes\n", bytes);
			if (nfq_handle_packet(nfq, packet, bytes))
				logMessage(LOG_WARNING, "Failed to handle packet: %s\n", strerror(errno));
		}
	}

	if (signalCaught)
	{
		const char *signalName = 
			signalCaught == SIGINT ? "SIGINT" : 
			signalCaught == SIGTERM ? "SIGTERM" : 
			signalCaught == SIGHUP ? "SIGHUP" : "??";

		logMessage(LOG_NOTICE, "Caught signal %s\n", signalName);
	}

	logMessage(LOG_DEBUG, "Destroying netfilter queue …\n");
	nfq_destroy_queue(queue);

	/* According to libnetfilter_queue's nfqnl_test.c example, nfq_unbind_pf(…)
	 * should NOT be called during clean up. */
	nfq_close(nfq);

	logMessage(LOG_NOTICE, "Exiting …\n");
	removePIDFile();
	destroyConfig();

	return exitCode;
}

static int inspectPacket(struct nfq_q_handle *queue, struct nfgenmsg *pktInfo, 
		struct nfq_data *pktData, void *userData)
{
	(void)pktInfo;
	(void)userData;

	uint8_t *packet;
	ssize_t size = nfq_get_payload(pktData, &packet);
	if (size < 0)
	{
		logMessage(LOG_WARNING, "Failed to retrieve packet from queue: %s\n",
				strerror(errno));
		return 1;
	}

	struct nfqnl_msg_packet_hdr *metaHeader = nfq_get_msg_packet_hdr(pktData);
	if (!packetIsComplete(packet, (size_t)size))
	{
		logMessage(LOG_INFO, "Dropping the packet because it is incomplete\n");
		return nfq_set_verdict(queue, ntohl(metaHeader->packet_id), NF_DROP, 0, NULL);
	}
	if (!packetIsDHCP(packet))
	{
		logMessage(LOG_DEBUG, "Ignoring non-DHCP packet\n");
		return nfq_set_verdict(queue, ntohl(metaHeader->packet_id), NF_ACCEPT, 0, NULL);
	}
	/* We do not have the logic needed to support fragmented packets: */
	if (ipv4_packetFragmented(&((const struct Packet *)packet)->ipHeader))
	{
		uint32_t verdict = config->fwdOnFail ? NF_ACCEPT : NF_DROP;
		if (config->fwdOnFail)
			logMessage(LOG_INFO, "Ingoring fragmented packet\n");
		else
			logMessage(LOG_INFO, "Dropping the packet because it is fragmented\n");
		
		return nfq_set_verdict(queue, ntohl(metaHeader->packet_id), verdict, 0, NULL);
	}
	if (config->debug)
		debugLogPacketHeader(packet, (size_t)size);

	logMessage(LOG_INFO, "Mangling packet\n");

	uint8_t *mangledData = NULL;
	size_t mangledDataSize = 0;
	enum MangleResult result = manglePacket(packet, (size_t)size, &mangledData,
			&mangledDataSize);
	if (result == Mangle_mallocFail)
	{
		logMessage(LOG_WARNING, "Failed to allocate memory for mangled packet\n");
		uint32_t verdict = config->fwdOnFail ? NF_ACCEPT : NF_DROP;
		return nfq_set_verdict(queue, ntohl(metaHeader->packet_id), verdict, 0, NULL);
	}
	else if (result == Mangle_optExists)
	{
		logMessage(LOG_INFO, "Dropping the packet because option already exists\n");
		return nfq_set_verdict(queue, ntohl(metaHeader->packet_id), NF_DROP, 0, NULL);
	}
	else if (result != Mangle_OK)
	{
		logMessage(LOG_ERR, "Internal error: unexpected return value from manglePacket(): %d\n",
				result);
		uint32_t verdict = config->fwdOnFail ? NF_ACCEPT : NF_DROP;
		return nfq_set_verdict(queue, ntohl(metaHeader->packet_id), verdict, 0, NULL);
	}

	if (config->debug)
		logMessage(LOG_DEBUG, "Sending mangled packet …\n");

	int res = nfq_set_verdict(queue, ntohl(metaHeader->packet_id), NF_ACCEPT, 
			mangledDataSize, mangledData);
	free(mangledData);
	return res;
}

static bool packetIsComplete(const uint8_t *data, size_t size)
{
	if (size < sizeof(struct IPv4Header))
		return false;

	const struct Packet *packet = (const struct Packet *)data;
	return packet->ipHeader.totalLen >= sizeof(*packet);
}

static bool packetIsDHCP(const uint8_t *data)
{
	const struct Packet *packet = (const struct Packet *)data;

	if (packet->ipHeader.protocol != IPPROTO_UDP)
		return false;

	uint16_t destPort = ntohs(packet->udpHeader.destPort);
	if (!(destPort == 67 || destPort == 68))
		return false;
	if (packet->udpHeader.length < sizeof(struct UDPHeader) + sizeof(struct BootP))
		return false;

	const struct BootP *dhcp = &packet->bootp;
	if (ntohl(dhcp->cookie) != DHCP_MAGIC_COOKIE)
		return false;

	return true;
}

static enum MangleResult manglePacket(const uint8_t *origData, size_t origDataSize,
		uint8_t **newData, size_t *newDataSize)
{
	const struct Packet *origPacket = (const struct Packet *)origData;
	size_t ipHdrSize = ipv4_headerLen(&origPacket->ipHeader);
	size_t udpHdrSize = sizeof(struct UDPHeader);
	size_t headersSize = ipHdrSize + udpHdrSize + sizeof(struct BootP);
	/* Allocate size for a new packet, slightly larger than needed in order to
	 * avoid reallocation.: */
	*newDataSize = origDataSize + config->dhcpOptsSize + 1; /* room for padding */
	size_t newPayloadSize = *newDataSize - ipHdrSize - udpHdrSize;
	/* Ensure that the DHCP packet (the BOOTP header and payload) is at least
	 * MIN_BOOTP_SIZE bytes long (as per the RFC 1542 requirement): */
	if (newPayloadSize < MIN_BOOTP_SIZE)
		*newDataSize += MIN_BOOTP_SIZE - newPayloadSize;

	*newData = malloc(*newDataSize);
	if (!*newData)
		return Mangle_mallocFail;

	/* Copy 'static' data (everything but the DHCP options) from original
	 * packet: */
	memcpy(*newData, origPacket, headersSize);
	enum MangleResult result = mangleOptions(origData, origDataSize, *newData, 
			newDataSize);
	if (result != Mangle_OK)
	{
		free(*newData);
		return result;
	}

	/* Recalculate actual size (and potential padding) after mangling options
	 * (the initially calculated size is possibly slightly too large, since it
	 * could not forsee how many bytes of DHCP options that was going to be
	 * removed; however, the header size fields need to be correct): */
	newPayloadSize = *newDataSize - ipHdrSize - udpHdrSize;
	size_t padding = (2 - (newPayloadSize % 2)) % 2;
	if (newPayloadSize < MIN_BOOTP_SIZE)
		padding = MIN_BOOTP_SIZE - newPayloadSize;

	newPayloadSize += padding;
	*newDataSize = ipHdrSize + udpHdrSize + newPayloadSize;

	struct Packet *newPacket = (struct Packet *)*newData;
	struct IPv4Header *ipHeader = &newPacket->ipHeader;
	ipHeader->totalLen = htons(*newDataSize);
	ipHeader->checksum = 0;
	ipHeader->checksum = ipv4_checksum(ipHeader);

	struct UDPHeader *udpHeader = &newPacket->udpHeader;
	udpHeader->length = htons(udpHdrSize + newPayloadSize);
	udpHeader->checksum = 0;

	if (padding && config->debug)
		logMessage(LOG_DEBUG, "Padding with %zu byte(s) to meet minimal BOOTP payload "
				"size …\n", padding);

	/* Pad to (at least) MIN_BOOTP_SIZE bytes: */
	for (size_t i = *newDataSize - padding; i < *newDataSize; ++i)
		(*newData)[i] = DHCPOPT_PAD;

	return Mangle_OK;
}

static enum MangleResult mangleOptions(const uint8_t *origData, size_t origDataSize,
		uint8_t *newData, size_t *newDataSize)
{
	/* Start with position of the first DHCP option: */
	size_t origOffset = offsetof(struct Packet, bootp) + sizeof(struct BootP);
	size_t newOffset = origOffset;
	size_t padCount = 0;
	while (origOffset < origDataSize)
	{
		const struct DHCPOption *option = (const struct DHCPOption *)(origData + origOffset);
		size_t optSize =
			option->code == DHCPOPT_PAD || option->code == DHCPOPT_END ? 1
			: sizeof(struct DHCPOption) + option->length;

		if (config->debug)
		{
			if (option->code == DHCPOPT_PAD)
				++padCount;
			else
			{
				if (padCount)
					logMessage(LOG_DEBUG, "Found %zu PAD options (removing)\n", padCount);

				debugLogOption(option);
				padCount = 0;
			}
		}

		if (option->code == DHCPOPT_END)
			break;
		/* If existing options are to be ignored and not removed, just copy
		 * them: */
		else if (config->ignoreExistOpt && !config->removeExistOpt)
		{
			if (config->debug)
				logMessage(LOG_DEBUG, " (copying)\n");

			memcpy(newData + newOffset, option, optSize);
			newOffset += optSize;
		}
		/* Otherwise we need to check whether one of the injected options are
		 * already present: */
		else
		{
			bool optFound = false;
			if (option->code != DHCPOPT_END)
				for (size_t i = 0; i < config->dhcpOptCodeCount; ++i)
					if (option->code == config->dhcpOptCodes[i])
					{
						optFound = true;
						break;
					}

			/* If the option already exists in original payload, but is not to be
			 * removed, and ignore command line option is not provided, drop
			 * packet: */
			if (optFound && !config->removeExistOpt && !config->ignoreExistOpt)
			{
				if (config->debug)
					logMessage(LOG_DEBUG, " (conflict)\n");

				return Mangle_optExists;
			}
			/* Copy option if it is not to be removed: */
			else if ((optFound && !config->removeExistOpt) || !optFound)
			{
				if (config->debug)
					logMessage(LOG_DEBUG, " (copying)\n");

				memcpy(newData + newOffset, option, optSize);
				newOffset += optSize;
			}
			else if (config->debug)
				logMessage(LOG_DEBUG, " (removing)\n");
		}
		origOffset += optSize;
	}

	if (config->debug)
		logMessage(LOG_DEBUG, "Injecting %zu option(s) …\n", config->dhcpOptCodeCount);

	/* Inject DHCP options: */
	for (size_t i = 0; i < config->dhcpOptsSize; ++i)
		newData[newOffset + i] = config->dhcpOpts[i];

	newOffset += config->dhcpOptsSize;

	if (config->debug)
		logMessage(LOG_DEBUG, "Inserting END option …\n");

	/* Finally insert the END option: */
	newData[newOffset++] = DHCPOPT_END;
	/* Update (reduce) packet size: */
	*newDataSize = newOffset;
	return Mangle_OK;
}

static void logMessage(int priority, const char *format, ...)
{
	if (priority == LOG_DEBUG && !config->debug)
		return;

	va_list args1, args2;
	va_start(args1, format);
	va_copy(args2, args1);

	if (config->foreground || !daemonised)
	{
		FILE *f = stderr;
		if (priority == LOG_NOTICE || priority == LOG_INFO || priority == LOG_DEBUG)
			f = stdout;
		
		vfprintf(f, format, args1);
	}
	va_end(args1);

	if (!config->foreground)
		vsyslog(priority, format, args2);

	va_end(args2);
}

static void simplifyProgramName(char *programName)
{
	char *simplifiedName = basename(programName);
	size_t len = strlen(simplifiedName);
	memmove(programName, simplifiedName, len);
	programName[len] = '\0';
}

static void writePID(void)
{
	if (!config->pidFile)
		return;

	pid_t pid = getpid();
	logMessage(LOG_DEBUG, "Writing PID %ld to %s …\n", (long)pid, config->pidFile);

	FILE *f = fopen(config->pidFile, "w");
	if (!f)
	{
		logMessage(LOG_ERR, "Failed to write PID to %s: %s\n", config->pidFile,
				strerror(errno));
		exit(EXIT_FAILURE);
	}
	fprintf(f, "%ld", (long)pid);
	fclose(f);
}

static void removePIDFile(void)
{
	if (config->pidFile)
	{
		logMessage(LOG_DEBUG, "Removing PID file %s …\n", config->pidFile);
		unlink(config->pidFile);
	}
}

static void destroyConfig(void)
{
	conf_destroy(config);
}

static void initSignalHandler(void)
{
	logMessage(LOG_DEBUG, "Initialising signal handler …\n");

	struct sigaction sigAction = { .sa_handler = &setEscapeMainLoopFlag };

	if (sigaction(SIGTERM, &sigAction, NULL) || sigaction(SIGINT, &sigAction, NULL) ||
			sigaction(SIGHUP, &sigAction, NULL))
	{
		logMessage(LOG_ERR, "Failed to initialise signal handler: %s\n", strerror(
					errno));
		exit(EXIT_FAILURE);
	}
}

static void setEscapeMainLoopFlag(int signal)
{
	signalCaught = signal;
	escapeMainLoop = true;
}

static void initLog(const char *programName)
{
	openlog(programName, 0, LOG_DAEMON);
	if (config->debug)
		setlogmask(LOG_UPTO(LOG_DEBUG));
	else
		setlogmask(LOG_UPTO(LOG_INFO));
}

static void debugLogOptions(void)
{
	if (!config->debug)
		return;

	logMessage(LOG_DEBUG, "%u DHCP option(s) to inject (total of %zu bytes): ",
			config->dhcpOptCodeCount, config->dhcpOptsSize);

	for (size_t i = 0; i < config->dhcpOptCodeCount; ++i)
	{
		uint8_t code = config->dhcpOptCodes[i];
		bool atEnd = i == config->dhcpOptCodeCount - 1;
		const char *delim = atEnd ? "\n" : ", ";
		logMessage(LOG_DEBUG, "0x%02X (%u)%s", code, code, delim);
	}
}

static void inspectOptions(void)
{
	size_t nonSpecialOptCount = 0;
	for (size_t i = 0; i < config->dhcpOptCodeCount; ++i)
	{
		uint8_t code = config->dhcpOptCodes[i];
		if (code != DHCPOPT_PAD && code != DHCPOPT_END)
			++nonSpecialOptCount;
	}

	if (!nonSpecialOptCount)
		logMessage(LOG_WARNING, "Warning: Only padding options added\n");
}

static void debugLogPacketHeader(const uint8_t *data, size_t size)
{
	const struct Packet *packet = (const struct Packet *)data;
	const uint8_t *mac = packet->bootp.clientHwAddr;
	struct IPAddr
	{
		uint8_t o1;
		uint8_t o2;
		uint8_t o3;
		uint8_t o4;
	} __attribute__((packed));

	const struct IPAddr *destIP = (const struct IPAddr *)&packet->ipHeader.destAddr; 

	logMessage(LOG_DEBUG, "Inspecting %zu-byte DHCP packet from "
			"%02X:%02X:%02X:%02X:%02X:%02X to %d.%d.%d.%d …\n",
			size,
			mac[0], mac[1], mac[2], mac[3], mac[4], mac[5],
			destIP->o1, destIP->o2, destIP->o3, destIP->o4
			);
}

static void debugLogOption(const struct DHCPOption *option)
{
	if (option->code == DHCPOPT_PAD)
		return;
	else if (option->code == DHCPOPT_END)
		logMessage(LOG_DEBUG,"Found END option %s\n", config->dhcpOptCodeCount ?
				"(removing)" : "(copying)");
	else if (option->code == DHCPOPT_TYPE && option->length == 1)
		logMessage(LOG_DEBUG, "Found option %hhu (0x%02hhX) (DHCP message type): %s",
				option->code, option->code, dhcp_msgTypeString(option->data[0]));
	else
	{
		/* String buffer for hex string (maximum DHCP option length (256) times
		 * three characters (two digits and a space)) */
		char optPayload[256 * 3];
		size_t i = 0;
		for (; i < option->length; ++i)
			sprintf(optPayload + 3*i, "%02X ", option->data[i]);

		/* Remove last space: */
		if (i)
			optPayload[3*i - 1] = '\0';

		logMessage(LOG_DEBUG, "Found option %hhu (0x%02hhX) with %' '3u-byte payload %s",
				option->code, option->code, option->length, optPayload);
	}
}
